### PR TITLE
Autorole!

### PR DESCRIPTION
**Pull request by nullpixel**
~~**_DO NOT MERGE:_ waiting for total days to be added courtesy of Doug.**~~

tested locally, works fine. fires every hour which can be tweaked. code is the best I could make it.

# Todo:

- [x] Auto de-role
- [x] Fix issue where user is demoted every time the function is rerun (caused by how we check for active users)
- [x] Log to #bot-log (coming soon:tm:)
- [x] Consecutive Days in database. (waiting for dougley on this)

# in this PR:

* auto role and derole
* show next rank in `!stats`, thx teaman
* Fail silently when chat voting fails because voting was closed. (#69)

### CODE REVIEW

**Comment by nullpixel** [comment]
[Users earn new roles automatically based on daily activity](https://trello.com/c/dAz57hnM/57-users-earn-new-roles-automatically-based-on-daily-activity)

File: Utils/orwell.js [hunk 1]
```diff
@@ -33,5 +34,20 @@ module.exports = {
...
+  roleUsers: (guild) => {
+    r.db("DFB").table("analytics").run().then((results) => {
+      results.forEach((row) => {
+        let totalDays = Object.keys(row.messages).length
+        let consecutiveDays = 7 // hardcoded until doug adds CD to schema
+        let member = guild.members.find(member => member.id === row.id)
+
+        Object.entries(roles).forEach(([key, role]) => {
+          if (totalDays && consecutiveDays >= role[0]) {
+            if (member) { member.assignRole(role[1]); console.log(`Gave ${member.id} ${key}`) }
```
Comment (LW001) [inline Utils/orwell.js]: Shouldn't it rather send the user a message or log it in bot-log? Thought it doesn't do console outputs

**Comment by nullpixel** [comment]
Almost done. See TODO in PR body.

**Comment by Dougley** [comment]
Style pls

**Comment by nullpixel** [comment]
got it.

**Comment by nullpixel** [comment]
I've done all the stuff I said I would for this - awaiting dougley to add Consecutive Days to the database

**Comment by Dougley** [comment]
Style pls

File: Utils/orwell.js [hunk 2]
```diff
@@ -33,5 +35,71 @@ module.exports = {
...
+              genlog.log(bot, bot.User, { 
+                message: `Added ${member.name}#${member.discriminator} to ${role.name}.`
+              })
+            })
+            return
+          }
+        })
+
+        if (!active) {
+          if (roleWeights === []) return; // has no roles :c
```
Comment (slice) [inline Utils/orwell.js]: you can't compare arrays like this in javascript lol

```js
> [] === []
< false
```

use `roleWeights.length`

**Review by slice — CHANGES_REQUESTED** [review]

File: Utils/orwell.js (hunk 2 above)
Comment (Arzte) [reply Utils/orwell.js]: fixed

File: roles.example.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  // Role ID :        { name, threshold, rank }
+  // Rank is a number from 1 to  5 where 1 is the lowest ranked role
+  // Deacy is the time in days to go from this role to custodian (lose a role every 7 days)
+  '273571650287566848': {name: 'Custodians', threshold: 3, rank: 1, decay: 7, message: `As thanks for being so active in our community, we've decided to award you the coveted Custodian role! This gives you access to the channel <#284796966641205249> and you can use a few new bot commands. Please check out <#268812893087203338> for more info.`},
+  '341246843558166529': {name: 'Record Keeper', threshold: 7, decay: 14, rank: 2},
+  '341520994160082945': {name: 'Book Keeper', threshold: 14, decay: 21, rank: 3},
+  '341521031246118913': {name: 'Librarian', threshold: 30, decay: 28, rank: 4},
+  '341521085637853197': {name: 'Vizier', threshold: 60, decay: 35, rank: 5},
+  '341521116675833866': {name: 'Grand Vizier', threshold: 120, decay: 42, rank: 6}
+}
```
Comment (Arzte) [inline roles.example.js:11]: Why do all of these need to be separate files?

Comment (Dougley) [reply roles.example.js:11]: Because I said so.

Comment (Arzte) [reply roles.example.js:11]: more files tho

**Comment by nullpixel** [comment]
@SteamingMutt any idea when your thing will be done?

File: Commands/Commands/queue.js [hunk 4]
```diff
@@ -370,6 +370,9 @@ commands.registerVote = {
                     cause: 'chat_vote',
                     message: (e.message !== undefined) ? e.message : JSON.stringify(e)
                   }, e)
+                } else if (e.statusCode === 422) {
```
Comment (nullpixel) [inline Commands/Commands/queue.js:373]: we can make this a switch case.

Comment (Arzte) [reply Commands/Commands/queue.js:373]: ... So why'd you merge it? lol

Comment (nullpixel) [reply Commands/Commands/queue.js:373]: because it's nothing with your pr, you followed the style of the old error catching, it was just a comment

Comment (Arzte) [reply Commands/Commands/queue.js:373]: And I was saying you could have done that on my pr, anyhow, we do this sort of thing on a lot of our commands, I wonder if there is something we can put in rather then these if statements everywhere? Of course This for example would need its own thing as its a "special" case.

Comment (nullpixel) [reply Commands/Commands/queue.js:373]: I think we should leave this as it is, then refactor all our error handling and huge if statements after, changing just this one doesn't make sense when we have a mess of them, along with messy callbacks and bad promise error catching

Comment (Arzte) [reply Commands/Commands/queue.js:373]: Maybe after this pr gets merged we look into refactoring the mess that's been created? (Specifically the huge if and error code)

**Review by Dougley — CHANGES_REQUESTED** [review]
The initial custodian role is on an approval basis in the admin-queue, however, this doesn't seem to be the case.

**Review by Dougley — APPROVED** [review]

**Pull request merged**